### ISSUE TITLE
Basic stacktrace augmentation

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#493]
After #333 we can do a basic [pluggable? opt-in or opt-out?] stacktrace augmentation on `throw` call-site.

The basic idea is to attempt to reflectively instantiate exception of the same type at `throw` call-site with original exception as a cause.

For example currently this code
```
val deferred = async {
  throw IllegalStateException("Some invariant failed")
}

try {
  deferred.await()
} catch (t: Throwable) {
  t.printStackTrace()
}
```

will print
```
java.lang.IllegalStateException: Some invariant failed
  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1$deferred$1.doResume(AsyncJvmTest. 
 kt:15)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)
```
after stacktrace augmentation it will be

```
java.lang.IllegalStateException: java.lang.IllegalStateException: Some invariant failed

  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1.doResume(AsyncJvmTest.kt:21)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)
  at kotlinx.coroutines.experimental.DispatchedContinuation.run(Dispatched.kt:13)
  at kotlinx.coroutines.experimental.EventLoopBase.processNextEvent(EventLoop.kt:132)
  ....
  at kotlinx.coroutines.experimental.AsyncJvmTest.simpleTest(AsyncJvmTest.kt:13) // <- WHERE EXCEPTION IS ACTUALLY THROWN
Caused by: java.lang.IllegalStateException: Some invariant failed
  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1$deferred$1.doResume(AsyncJvmTest.kt:15)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)

```

FJP for FJ-tasks does exactly the same: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/ForkJoinTask.java#l582
